packages/workspaces-extension/src/commands.ts

```diff
@@ -0,0 +1,585 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IRouter,
+  JupyterFrontEnd,
+  JupyterFrontEndPlugin
+} from '@jupyterlab/application';
+import {
+  Dialog,
+  InputDialog,
+  IWindowResolver,
+  showDialog
+} from '@jupyterlab/apputils';
+
+import { URLExt } from '@jupyterlab/coreutils';
+
+import { FileDialog, IDefaultFileBrowser } from '@jupyterlab/filebrowser';
+import { Contents, Workspace } from '@jupyterlab/services';
+import { IStateDB } from '@jupyterlab/statedb';
+import { ITranslator, nullTranslator } from '@jupyterlab/translation';
+import { ICommandPalette } from '@jupyterlab/apputils';
+import { IWorkspaceCommands, IWorkspacesModel } from '@jupyterlab/workspaces';
+
+namespace CommandIDs {
+  /**
+   * Open a workspace by identifier. When no identifier is supplied, falls back to a simple choice dialog.
+   */
+  export const open = 'workspace-ui:open';
+  /**
+   * Save the current workspace.
+   */
+  export const save = 'workspace-ui:save';
+  /**
+   * Save the current workspace under a different path.
+   */
+  export const saveAs = 'workspace-ui:save-as';
+  /**
+   * Create a new workspace.
+   */
+  export const createNew = 'workspace-ui:create-new';
+  /**
+   * Delete a workspace.
+   */
+  export const deleteWorkspace = 'workspace-ui:delete';
+  /**
+   * Clone a given workspace.
+   */
+  export const clone = 'workspace-ui:clone';
+  /**
+   * Rename a workspace.
+   */
+  export const rename = 'workspace-ui:rename';
+  /**
+   * Reset a workspace.
+   */
+  export const reset = 'workspace-ui:reset';
+  /**
+   * Import a workspace.
+   */
+  export const importWorkspace = 'workspace-ui:import';
+  /**
+   * Export a workspace.
+   */
+  export const exportWorkspace = 'workspace-ui:export';
+}
+
+const WORKSPACE_NAME = 'jupyterlab-workspace';
+const WORKSPACE_EXT = '.' + WORKSPACE_NAME;
+const LAST_SAVE_ID = 'workspace-ui:lastSave';
+
+export const WORKSPACE_ITEM_CLASS = 'jp-mod-workspace';
+
+/**
+ * The workspace commands
+ */
+export const commandsPlugin: JupyterFrontEndPlugin<IWorkspaceCommands> = {
+  id: '@jupyterlab/workspaces:commands',
+  description: 'Add workspace commands.',
+  autoStart: true,
+  requires: [
+    IWorkspacesModel,
+    IDefaultFileBrowser,
+    IWindowResolver,
+    IStateDB,
+    ITranslator,
+    JupyterFrontEnd.IPaths
+  ],
+  provides: IWorkspaceCommands,
+  optional: [IRouter, ICommandPalette],
+  activate: (
+    app: JupyterFrontEnd,
+    model: IWorkspacesModel,
+    fileBrowser: IDefaultFileBrowser,
+    resolver: IWindowResolver,
+    state: IStateDB,
+    translator: ITranslator,
+    paths: JupyterFrontEnd.IPaths,
+    router: IRouter | null,
+    palette: ICommandPalette | null
+  ): IWorkspaceCommands => {
+    const trans = translator.load('jupyterlab');
+
+    const namingHintLabel = trans.__(
+      'Naming the workspace will create a unique URL. Please name your workspace without special characters.'
```

Comment:
`jupyterlab-server` explicitly allows spaces and does not list any restrictions on manager level (it does use [`slugify`](https://github.com/jupyterlab/jupyterlab_server/blob/91a081672f05d6611b7d7a9f832695e863403c1d/jupyterlab_server/workspaces_handler.py#L66-L95) to remove invalid characters when creating file name, but it has no bearing on workspace resolution from URL).

There are restrictions on the routing/handler level though:

```
MASTER_URL_PATTERN = (
    r"/(?P<mode>{}|doc)(?P<workspace>/workspaces/[a-zA-Z0-9\-\_]+)?(?P<tree>/tree/.*)?"
)
```

[`MASTER_URL_PATTERN`](https://github.com/jupyterlab/jupyterlab_server/blob/91a081672f05d6611b7d7a9f832695e863403c1d/jupyterlab_server/handlers.py#L34) only allows `[a-zA-Z0-9\-\_]+`.

Interestingly there is no such a limitation on the Workspaces Manager API:

```
workspace_api_path = ujoin(extension_app.workspaces_api_url, "(?P<space_name>.+)")
        handlers.append((workspace_api_path, WorkspacesHandler, workspaces_config))
```

which takes basically any combination of characters (`.+`) - defined [here](
https://github.com/jupyterlab/jupyterlab_server/blob/91a081672f05d6611b7d7a9f832695e863403c1d/jupyterlab_server/handlers.py#L266).

To be on the safe side I will update the instructions and add validation for `[a-zA-Z0-9\-\_]+`